support-lib/cpp/SharedFuture.hpp

```diff
@@ -0,0 +1,161 @@
+        Func transform) const {
+        auto cpy = SharedFuture(*this); // retain copy during coroutine suspension
+        co_await cpy.waitIgnoringExceptions();
+        co_return transform(cpy);
+    }
+
+    // Overload for T = void or `transform` takes no arugment.
+    template<typename Func, typename = std::enable_if_t<!std::is_invocable_v<Func, const SharedFuture<T>&>>>
+    SharedFuture<std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<Func>>>> then(Func transform) const {
+        co_await SharedFuture(*this); // retain copy during coroutine suspension
+        co_return transform();
+    }
+
+    // -- coroutine support implementation only; not intended externally --
+
+    bool await_ready() const {
+        std::scoped_lock lock(_sharedStates->mutex);
+        return _sharedStates->storedValue.has_value();
+    }
+
+    decltype(auto) await_resume() const {
+        if (!*_sharedStates->storedValue) {
+            std::rethrow_exception(_sharedStates->storedValue->error());
+        }
+        if constexpr (!std::is_void_v<T>) {
+            return const_cast<const T &>(_sharedStates->storedValue->value());
+        }
+    }
+
+    bool await_suspend(detail::CoroutineHandle<> h) const;
+
+    struct Promise : public Future<T>::promise_type {
+        SharedFuture<T> get_return_object() noexcept {
+            return SharedFuture(Future<T>::promise_type::get_return_object());
+        }
+    };
+    using promise_type = Promise;
+
+private:
+    Future<void> waitIgnoringExceptions() const {
+        try {
+            co_await *this;
+        } catch (...) {
+            // Ignore exceptions.
+        }
+    }
+
+    struct SharedStates {
+        std::recursive_mutex mutex;
+        std::optional<djinni::expected<T, std::exception_ptr>> storedValue = std::nullopt;
+        std::vector<detail::CoroutineHandle<>> coroutineHandles;
+    };
+    // Use a shared_ptr to allow copying SharedFuture.
+    std::shared_ptr<SharedStates> _sharedStates = std::make_shared<SharedStates>();
+};
+
+// CTAD deduction guide to construct from Future directly.
+template<typename T>
+SharedFuture(Future<T>&&) -> SharedFuture<T>;
+
+// ------------------ Implementation ------------------
+
+template<typename T>
+SharedFuture<T>::SharedFuture(Future<T>&& future) {
+    // `future` will invoke all continuations when it is ready.
+    future.then([sharedStates = _sharedStates](auto futureResult) {
+        std::vector toCall = [&] {
+            std::scoped_lock lock(sharedStates->mutex);
+            try {
+                if constexpr (std::is_void_v<T>) {
+                    futureResult.get();
+                    sharedStates->storedValue.emplace();
+                } else {
+                    sharedStates->storedValue = futureResult.get();
+                }
+            } catch (const std::exception& e) {
+                sharedStates->storedValue = make_unexpected(std::make_exception_ptr(e));
+            }
```

Comment:
```suggestion
            } catch (...) {
                sharedStates->storedValue = make_unexpected(std::current_exception());
            }
```

The current implementation runs into this issue: https://stackoverflow.com/questions/31925998/weird-behavior-using-stdmake-exception-ptr

Reply:
good catch